packages/php-wasm/universal/src/lib/php-request-handler.ts

```diff
@@ -306,14 +319,89 @@ export class PHPRequestHandler {
+
+		if (primaryPhp.isDir(fsPath)) {
+			// Ensure directory URIs have a trailing slash. Otherwise,
+			// relative URIs in index.php or index.html files are relative
+			// to the next directory up.
+			//
+			// Example:
+			// For a request to "/wp-admin", the relative link "edit.php"
+			// resolves to "/edit.php" rather than "/wp-admin/edit.php".
+			//
+			// This is correct behavior for the browser:
+			// https://www.rfc-editor.org/rfc/rfc3986#section-5.2.3
+			//
+			// But the intent for `/wp-admin/index.php` is that its relative
+			// URIs are relative to `/wp-admin/`.
+			//
+			// In fact, WordPress also redirects like this when given a chance.
+			// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/canonical.php#L696
+			// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/canonical.php#L1036-L1045
+			// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/link-template.php#L3558
+			if (!fsPath.endsWith('/')) {
+				return new PHPResponse(
+					301,
+					{ Location: [`${requestedUrl.pathname}/`] },
+					new Uint8Array(0)
+				);
+			}
+
+			// We can only satisfy requests for directories with a default file
+			// so let's first resolve to a default path when available.
+			const defaultFilePath = ['index.php', 'index.html']
+				.map((defaultFilename) => joinPaths(fsPath, defaultFilename))
+				.find((possibleDefaultPath) =>
+					primaryPhp.isFile(possibleDefaultPath)
+				);
+
+			if (defaultFilePath) {
+				fsPath = defaultFilePath;
+			}
+		}
+
+		if (fsPath.endsWith('.php')) {
+			if (primaryPhp.isFile(fsPath)) {
+				const effectiveRequest: PHPRequest = {
+					...request,
+					url: joinPaths(this.#ABSOLUTE_URL, fsPath),
+				};
+				return this.#spawnPHPAndDispatchRequest(
+					effectiveRequest,
+					requestedUrl
+				);
+			}
+		} else {
+			if (primaryPhp.isFile(fsPath)) {
+				return this.#serveStaticFile(primaryPhp, fsPath);
+			} else if (
+				// Make sure fsPath doesn't describe any other entity on the filesystem
+				!primaryPhp.fileExists(fsPath) &&
+				this.isKnownRemoteAssetPath(normalizedRequestedPath)
+			) {
+				// This path is listed as a remote asset. Mark it as a static file
+				// so the service worker knows it can issue a real fetch() to the server.
+				return new PHPResponse(
+					404,
+					{ 'x-file-type': ['static'] },
+					new TextEncoder().encode('404 File not found')
+				);
+			}
 		}
-		return this.#spawnPHPAndDispatchRequest(request, requestedUrl);
+
+		// TODO: Can we adjust this default to reduce coupling between the request handler and WP?
+		// Delegate unresolved requests to WordPress. This makes WP magic possible,
+		// like pretty permalinks and dynamically generated sitemaps.
+		const wpDefaultPath = joinPaths(this.#DOCROOT, 'index.php');
```

Comment:
This may be fine for now, but over time, I think we may appreciate if PHPRequestHandler is decoupled from WordPress. AFAICT, it mostly is prior to this PR which changes request handling to handle unknown URIs with `/index.php`.

So I left a couple a few duplicate TODOs scattered in this module like:
```
// TODO: Can we adjust this default to reduce coupling between the request handler and WP?
```

It may be that we just remove these TODOs and accept the coupling for now, but I think it's worth bringing up.

Reply:
> changes request handling to handle unknown URIs with /index.php

Apache has an `ErrorDocument` directive that you can use to handle 404 errors using a specific PHP file. We could introduce a similar `ErrorDocument` option here and:

* Default to `joinPaths(this.#DOCROOT, 'index.php')`
* Return 404 when it's explicitly set to null

Reply:
> Apache has an ErrorDocument directive that you can use to handle 404 errors using a specific PHP file.

PHPRequestHandler now supports an optional `getFileNotFoundAction()` callback for this, and `bootWordPress()` has been updated to provide its own WordPress-friendly file-not-found handling.